File: shared/src/mcp-client.ts
import fetch, { type RequestInit, type Response } from 'node-fetch';

export interface MCPToolCall {
  service: string;
  tool: string;
  args: Record<string, unknown>;
}

export interface MCPResult {
  success: boolean;
  data?: unknown;
  error?: string;
}

export interface MCPClientOptions {
  gatewayUrl?: string;
  accessToken?: string;
  timeoutMs?: number;
  fetchImpl?: (url: string, init?: RequestInit) => Promise<Response>;
}

interface JsonRpcResponse {
  jsonrpc?: string;
  result?: unknown;
  error?: { code?: number; message?: string };
}

/**
 * Canonical gateway client. It never executes tools locally and never converts
 * transport failures into successful MCP results.
 */
export class MCPClient {
  private readonly gatewayUrl: string;
  private readonly accessToken: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: (url: string, init?: RequestInit) => Promise<Response>;
  private requestId = 0;

  constructor(options: MCPClientOptions = {}) {
    this.gatewayUrl = (options.gatewayUrl ?? process.env.MCP_GATEWAY_URL ?? '').replace(/\/$/, '');
    this.accessToken = options.accessToken ?? process.env.MCP_GATEWAY_TOKEN ?? '';
    this.timeoutMs = options.timeoutMs ?? 30_000;
    this.fetchImpl = options.fetchImpl ?? fetch;
  }

  async callQATool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('qa-mcp', tool, args);
  }

  async callTestTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('test-mcp', tool, args);
  }

  async callDocsTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('docs-mcp', tool, args);
  }

  async callDeployTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('deploy-mcp', tool, args);
  }

  async callSessionTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('session-mcp', tool, args);
  }

  async callInfraTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('infra-mcp', tool, args);
  }

  async callGovernanceTool(tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    return this.callTool('ai-governance-mcp', tool, args);
  }

  async callTool(service: string, tool: string, args: Record<string, unknown>): Promise<MCPResult> {
    if (!this.gatewayUrl) {
      return { success: false, error: 'MCP_GATEWAY_URL is required' };
    }
    if (!this.accessToken) {
      return { success: false, error: 'MCP_GATEWAY_TOKEN is required' };
    }
    if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(service)) {
      return { success: false, error: 'Invalid MCP service id' };
    }
    if (!/^[a-z][a-z0-9_-]*$/.test(tool)) {
      return { success: false, error: 'Invalid MCP tool name' };
    }
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(
        `${this.gatewayUrl}/mcp/${encodeURIComponent(service)}/tools/call`,
        {
          method: 'POST',
          headers: {
            Authorization: `Bearer ${this.accessToken}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({ id: ++this.requestId, name: tool, arguments: args }),
          signal: controller.signal,
        },
      );
      if (!response.ok) {
        return { success: false, error: `MCP gateway returned HTTP ${response.status}` };
      }
      let payload: JsonRpcResponse;
      try {
        payload = (await response.json()) as JsonRpcResponse;
      } catch {
        return { success: false, error: 'MCP gateway returned invalid JSON' };
      }
      if (payload.error) {
        return { success: false, error: payload.error.message ?? 'MCP tool returned an error' };
      }
      if (!Object.prototype.hasOwnProperty.call(payload, 'result')) {
        return { success: false, error: 'MCP gateway response has no result' };
      }
      return { success: true, data: payload.result };
    } catch (error) {
      const message = error instanceof Error && error.name === 'AbortError'
        ? 'MCP gateway request timed out'
        : 'MCP gateway request failed';
      return { success: false, error: message };
    } finally {
      clearTimeout(timer);
    }
  }

  async validateWithQA(docPath: string): Promise<boolean> {
    return (await this.callQATool('run_linter', { repo_path: docPath })).success;
  }

  async createTestPlan(feature: string): Promise<unknown> {
    return (await this.callTestTool('create_test_plan', {
      title: `Test Plan: ${feature}`,
      scope: `Testing ${feature}`,
    })).data;
  }

  async generateDocumentation(title: string, vars: Record<string, string>): Promise<unknown> {
    return (await this.callDocsTool('generate_doc', { template_name: title, variables: vars })).data;
  }

  async commitBacklog(files: Array<{ path: string; content: string }>): Promise<unknown> {
    return (await this.callDeployTool('commit_files', {
      repo: 'platform-devs', branch: 'main', message: 'docs: backlog and tasks created', files,
    })).data;
  }

  async recordArtifact(sessionId: string, type: string, content: string): Promise<MCPResult> {
    return this.callSessionTool('add_artifact', {
      session_id: sessionId, artifact_type: type, content,
    });
  }
}

let defaultClient: MCPClient | undefined;

export function getMCPClient(): MCPClient {
  defaultClient ??= new MCPClient();
  return defaultClient;
}

/** Backwards-compatible singleton; configuration is validated on first call. */
export const mcpClient = new MCPClient();

File: shared/mcp_client.py
"""Fail-closed client for cross-MCP calls through the canonical gateway."""

from __future__ import annotations

import os
import re
from typing import Any

import httpx

_SERVICE_ID = re.compile(r"^[a-z0-9]+(?:-[a-z0-9]+)*$")
_TOOL_NAME = re.compile(r"^[a-z][a-z0-9_-]*$")


class MCPClientError(RuntimeError):
    """A gateway transport or protocol failure."""


class MCPToolError(MCPClientError):
    """A tool returned a canonical JSON-RPC error."""


class MCPHttpClient:
    """Call governed tools without direct provider URLs or local fallbacks."""

    def __init__(
        self,
        gateway_url: str | None = None,
        access_token: str | None = None,
        *,
        timeout: float = 30.0,
        client: httpx.Client | None = None,
    ) -> None:
        self.gateway_url = (gateway_url or os.environ.get("MCP_GATEWAY_URL", "")).rstrip("/")
        self._access_token = access_token or os.environ.get("MCP_GATEWAY_TOKEN", "")
        if not self.gateway_url:
            raise MCPClientError("MCP_GATEWAY_URL is required")
        if not self._access_token:
            raise MCPClientError("MCP_GATEWAY_TOKEN is required")
        self._owns_client = client is None
        self.client = client or httpx.Client(timeout=timeout)
        self._request_id = 0

    def call(self, service: str, tool: str, args: dict[str, Any] | None = None) -> Any:
        if not _SERVICE_ID.fullmatch(service):
            raise MCPClientError("invalid MCP service id")
        if not _TOOL_NAME.fullmatch(tool):
            raise MCPClientError("invalid MCP tool name")
        self._request_id += 1
        try:
            response = self.client.post(
                f"{self.gateway_url}/mcp/{service}/tools/call",
                headers={"Authorization": f"Bearer {self._access_token}"},
                json={"id": self._request_id, "name": tool, "arguments": args or {}},
            )
            response.raise_for_status()
            payload = response.json()
        except httpx.TimeoutException as exc:
            raise MCPClientError("MCP gateway request timed out") from exc
        except httpx.HTTPStatusError as exc:
            raise MCPClientError(f"MCP gateway returned HTTP {exc.response.status_code}") from exc
        except (httpx.HTTPError, ValueError) as exc:
            raise MCPClientError("MCP gateway request failed") from exc
        if not isinstance(payload, dict):
            raise MCPClientError("MCP gateway returned a non-object response")
        error = payload.get("error")
        if isinstance(error, dict):
            raise MCPToolError(str(error.get("message") or "MCP tool returned an error"))
        if "result" not in payload:
            raise MCPClientError("MCP gateway response has no result")
        return payload["result"]

    def close(self) -> None:
        if self._owns_client:
            self.client.close()

    def __enter__(self) -> "MCPHttpClient":
        return self

    def __exit__(self, *_: object) -> None:
        self.close()


_client: MCPHttpClient | None = None


def get_mcp_client() -> MCPHttpClient:
    global _client
    if _client is None:
        _client = MCPHttpClient()
    return _client
